File: reservation-missing-order-id/node/trace-reservation-orders.js
/**
 * Trace Medusa reservations back to the orders they belong to.
 *
 * Medusa v2 deliberately decouples the Inventory module from the Order module.
 * ReservationItem stores only a bare line_item_id string, not a real relation, and
 * there is no module link between ReservationItem and the Order module, so the
 * Admin API and dashboard can never show which order a reservation is for. This
 * lists reservations and orders, builds a line item to order lookup that stands in
 * for the Order module's own OrderItem join, and reports every reservation as
 * traced, orphaned, or not order backed. The only write is optional enrichment:
 * stamping the resolved order id into a traced reservation's own metadata.
 * Run on a schedule, or on demand. Safe to run again and again.
 *
 * Guide: https://www.allanninal.dev/medusa/reservation-missing-order-id/
 */
import { pathToFileURL } from "node:url";
import { writeFileSync } from "node:fs";

const BACKEND_URL = process.env.MEDUSA_BACKEND_URL || "http://localhost:9000";
const ADMIN_EMAIL = process.env.MEDUSA_ADMIN_EMAIL || "admin@example.com";
const ADMIN_PASSWORD = process.env.MEDUSA_ADMIN_PASSWORD || "supersecret";
const REPORT_PATH = process.env.REPORT_PATH || "reservation_trace_report.csv";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

/**
 * Pure decision function. No I/O.
 *
 * Build a lookup: order line-item id -> order id, by walking each order's items[]
 * (this stands in for the real OrderItem join: Order 1--* OrderItem *--1 OrderLineItem).
 *
 * @param {{ id: string, line_item_id: string | null, inventory_item_id: string, quantity: number }[]} reservations
 * @param {{ id: string, items: { id: string }[] }[]} orders
 * @returns {{ reservation_id: string, order_id: string | null, status: "traced" | "orphaned_line_item" | "no_line_item" }[]}
 */
export function traceReservationsToOrders(reservations, orders) {
  const lineItemToOrder = new Map();
  for (const order of orders) {
    for (const item of order.items) {
      lineItemToOrder.set(item.id, order.id);
    }
  }

  return reservations.map((r) => {
    if (!r.line_item_id) {
      return { reservation_id: r.id, order_id: null, status: "no_line_item" };
    }
    const orderId = lineItemToOrder.get(r.line_item_id);
    if (!orderId) {
      return { reservation_id: r.id, order_id: null, status: "orphaned_line_item" };
    }
    return { reservation_id: r.id, order_id: orderId, status: "traced" };
  });
}

async function getAdminToken() {
  const res = await fetch(`${BACKEND_URL}/auth/user/emailpass`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ email: ADMIN_EMAIL, password: ADMIN_PASSWORD }),
  });
  if (!res.ok) throw new Error(`Medusa auth ${res.status}`);
  const body = await res.json();
  return body.token;
}

async function adminGet(token, path, params = {}) {
  const url = new URL(`${BACKEND_URL}${path}`);
  for (const [key, value] of Object.entries(params)) {
    url.searchParams.set(key, value);
  }
  const res = await fetch(url, { headers: { Authorization: `Bearer ${token}` } });
  if (!res.ok) throw new Error(`Medusa ${res.status} on GET ${path}`);
  return res.json();
}

async function adminPost(token, path, body) {
  const res = await fetch(`${BACKEND_URL}${path}`, {
    method: "POST",
    headers: { Authorization: `Bearer ${token}`, "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(`Medusa ${res.status} on POST ${path}`);
  return res.json();
}

async function listReservations(token) {
  const reservations = [];
  let offset = 0;
  const limit = 200;
  while (true) {
    const data = await adminGet(token, "/admin/reservations", {
      fields: "id,inventory_item_id,location_id,quantity,line_item_id,created_at",
      limit,
      offset,
    });
    reservations.push(...data.reservations);
    offset += limit;
    if (offset >= data.count) return reservations;
  }
}

async function listOrders(token) {
  const orders = [];
  let offset = 0;
  const limit = 200;
  while (true) {
    const data = await adminGet(token, "/admin/orders", {
      fields: "id,*items",
      limit,
      offset,
    });
    orders.push(...data.orders);
    offset += limit;
    if (offset >= data.count) return orders;
  }
}

async function stampResolvedOrderId(token, reservationId, orderId) {
  return adminPost(token, `/admin/reservations/${reservationId}`, {
    metadata: { resolved_order_id: orderId },
  });
}

function writeReport(path, rows) {
  const header = "reservation_id,inventory_item_id,order_id,status";
  const lines = rows.map((r) =>
    [r.reservation_id, r.inventory_item_id, r.order_id || "", r.status].join(",")
  );
  writeFileSync(path, [header, ...lines].join("\n") + "\n");
}

export async function run() {
  const token = await getAdminToken();
  const reservations = await listReservations(token);
  const orders = await listOrders(token);
  const trace = traceReservationsToOrders(reservations, orders);

  const byId = new Map(reservations.map((r) => [r.id, r]));
  const reportRows = [];
  let enriched = 0;
  let orphaned = 0;
  let noLineItem = 0;

  for (const result of trace) {
    const reservation = byId.get(result.reservation_id);
    reportRows.push({
      reservation_id: result.reservation_id,
      inventory_item_id: reservation.inventory_item_id,
      order_id: result.order_id || "",
      status: result.status,
    });

    if (result.status === "traced") {
      console.log(
        `Reservation ${result.reservation_id} traced to order ${result.order_id}. ${DRY_RUN ? "would stamp metadata" : "stamping metadata"}`
      );
      if (!DRY_RUN) await stampResolvedOrderId(token, result.reservation_id, result.order_id);
      enriched++;
    } else if (result.status === "orphaned_line_item") {
      console.warn(`Reservation ${result.reservation_id} has an orphaned line_item_id, flagged for manual review.`);
      orphaned++;
    } else {
      noLineItem++;
    }
  }

  writeReport(REPORT_PATH, reportRows);
  console.log(
    `Done. ${enriched} traced (${DRY_RUN ? "would enrich" : "enriched"}), ${orphaned} orphaned, ${noLineItem} not order backed. Report written to ${REPORT_PATH}.`
  );
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: reservation-missing-order-id/python/trace_reservation_orders.py
"""Trace Medusa reservations back to the orders they belong to.

Medusa v2 deliberately decouples the Inventory module from the Order module.
ReservationItem stores only a bare line_item_id string, not a real relation, and
there is no module link between ReservationItem and the Order module, so the
Admin API and dashboard can never show which order a reservation is for. This
lists reservations and orders, builds a line item to order lookup that stands in
for the Order module's own OrderItem join, and reports every reservation as
traced, orphaned, or not order backed. The only write is optional enrichment:
stamping the resolved order id into a traced reservation's own metadata.
Run on a schedule, or on demand. Safe to run again and again.

Guide: https://www.allanninal.dev/medusa/reservation-missing-order-id/
"""
import os
import csv
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("trace_reservation_orders")

BACKEND_URL = os.environ["MEDUSA_BACKEND_URL"]
ADMIN_EMAIL = os.environ["MEDUSA_ADMIN_EMAIL"]
ADMIN_PASSWORD = os.environ["MEDUSA_ADMIN_PASSWORD"]
REPORT_PATH = os.environ.get("REPORT_PATH", "reservation_trace_report.csv")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"


def get_admin_token():
    r = requests.post(
        f"{BACKEND_URL}/auth/user/emailpass",
        json={"email": ADMIN_EMAIL, "password": ADMIN_PASSWORD},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()["token"]


def admin_get(token, path, params=None):
    r = requests.get(
        f"{BACKEND_URL}{path}",
        headers={"Authorization": f"Bearer {token}"},
        params=params or {},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def admin_post(token, path, body):
    r = requests.post(
        f"{BACKEND_URL}{path}",
        headers={"Authorization": f"Bearer {token}"},
        json=body,
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def trace_reservations_to_orders(reservations, orders):
    """Pure decision function. No I/O.

    reservations: [{"id": str, "line_item_id": str | None, "inventory_item_id": str, "quantity": int}]
    orders: [{"id": str, "items": [{"id": str}]}]

    Returns [{"reservation_id": str, "order_id": str | None,
              "status": "traced" | "orphaned_line_item" | "no_line_item"}]

    Build a lookup: order line-item id -> order id, by walking each order's items[]
    (this stands in for the real OrderItem join: Order 1--* OrderItem *--1 OrderLineItem).
    """
    line_item_to_order = {}
    for order in orders:
        for item in order.get("items") or []:
            line_item_to_order[item["id"]] = order["id"]

    results = []
    for r in reservations:
        line_item_id = r.get("line_item_id")
        if not line_item_id:
            results.append({"reservation_id": r["id"], "order_id": None, "status": "no_line_item"})
            continue
        order_id = line_item_to_order.get(line_item_id)
        if not order_id:
            results.append({"reservation_id": r["id"], "order_id": None, "status": "orphaned_line_item"})
            continue
        results.append({"reservation_id": r["id"], "order_id": order_id, "status": "traced"})
    return results


def list_reservations(token):
    reservations = []
    offset = 0
    limit = 200
    while True:
        data = admin_get(token, "/admin/reservations", {
            "fields": "id,inventory_item_id,location_id,quantity,line_item_id,created_at",
            "limit": limit,
            "offset": offset,
        })
        reservations.extend(data["reservations"])
        offset += limit
        if offset >= data["count"]:
            return reservations


def list_orders(token):
    orders = []
    offset = 0
    limit = 200
    while True:
        data = admin_get(token, "/admin/orders", {
            "fields": "id,*items",
            "limit": limit,
            "offset": offset,
        })
        orders.extend(data["orders"])
        offset += limit
        if offset >= data["count"]:
            return orders


def stamp_resolved_order_id(token, reservation_id, order_id):
    return admin_post(token, f"/admin/reservations/{reservation_id}", {
        "metadata": {"resolved_order_id": order_id},
    })


def write_report(path, rows):
    with open(path, "w", newline="") as f:
        writer = csv.DictWriter(f, fieldnames=["reservation_id", "inventory_item_id", "order_id", "status"])
        writer.writeheader()
        for row in rows:
            writer.writerow(row)


def run():
    token = get_admin_token()
    reservations = list_reservations(token)
    orders = list_orders(token)
    trace = trace_reservations_to_orders(reservations, orders)

    by_id = {r["id"]: r for r in reservations}
    report_rows = []
    enriched = 0
    orphaned = 0
    no_line_item = 0

    for result in trace:
        reservation = by_id[result["reservation_id"]]
        report_rows.append({
            "reservation_id": result["reservation_id"],
            "inventory_item_id": reservation["inventory_item_id"],
            "order_id": result["order_id"] or "",
            "status": result["status"],
        })

        if result["status"] == "traced":
            log.info(
                "Reservation %s traced to order %s. %s",
                result["reservation_id"], result["order_id"],
                "would stamp metadata" if DRY_RUN else "stamping metadata",
            )
            if not DRY_RUN:
                stamp_resolved_order_id(token, result["reservation_id"], result["order_id"])
            enriched += 1
        elif result["status"] == "orphaned_line_item":
            log.warning(
                "Reservation %s has an orphaned line_item_id, flagged for manual review.",
                result["reservation_id"],
            )
            orphaned += 1
        else:
            no_line_item += 1

    write_report(REPORT_PATH, report_rows)
    log.info(
        "Done. %d traced (%s), %d orphaned, %d not order backed. Report written to %s.",
        enriched, "enriched" if not DRY_RUN else "would enrich", orphaned, no_line_item, REPORT_PATH,
    )


if __name__ == "__main__":
    run()
